op-node/rollup/derive/batch_queue.go

```diff
@@ -53,99 +59,152 @@ type BatchQueue struct {
 	// length of l1Blocks never exceeds SequencerWindowSize
 	l1Blocks []eth.L1BlockRef
 
-	// batches in order of when we've first seen them
-	batches []*BatchWithL1InclusionBlock
-
 	// nextSpan is cached SingularBatches derived from SpanBatch
 	nextSpan []*SingularBatch
 
 	l2 SafeBlockFetcher
 }
 
-// NewBatchQueue creates a BatchQueue, which should be Reset(origin) before use.
-func NewBatchQueue(log log.Logger, cfg *rollup.Config, prev NextBatchProvider, l2 SafeBlockFetcher) *BatchQueue {
-	return &BatchQueue{
+func newBaseBatchStage(log log.Logger, cfg *rollup.Config, prev NextBatchProvider, l2 SafeBlockFetcher) baseBatchStage {
+	return baseBatchStage{
 		log:    log,
 		config: cfg,
 		prev:   prev,
 		l2:     l2,
 	}
 }
 
-func (bq *BatchQueue) Origin() eth.L1BlockRef {
-	return bq.prev.Origin()
+func (bs *baseBatchStage) base() *baseBatchStage {
+	return bs
+}
+
+func (bs *baseBatchStage) Log() log.Logger {
+	if len(bs.l1Blocks) == 0 {
+		return bs.log.New("origin", bs.origin.ID())
+	} else {
+		return bs.log.New("origin", bs.origin.ID(), "epoch", bs.l1Blocks[0])
+	}
+}
+
+type SingularBatchProvider interface {
+	ResettableStage
+	NextBatch(context.Context, eth.L2BlockRef) (*SingularBatch, bool, error)
+}
+
+// BatchQueue contains a set of batches for every L1 block.
+// L1 blocks are contiguous and this does not support reorgs.
+type BatchQueue struct {
+	baseBatchStage
+
+	// batches in order of when we've first seen them
+	batches []*BatchWithL1InclusionBlock
+}
+
+var _ SingularBatchProvider = (*BatchQueue)(nil)
+
+// NewBatchQueue creates a BatchQueue, which should be Reset(origin) before use.
+func NewBatchQueue(log log.Logger, cfg *rollup.Config, prev NextBatchProvider, l2 SafeBlockFetcher) *BatchQueue {
+	return &BatchQueue{baseBatchStage: newBaseBatchStage(log, cfg, prev, l2)}
+}
```

Comment:
Agree, it happened organically. I'll move it around later (https://github.com/ethereum-optimism/optimism/issues/12567) to avoid conflicts with stacked PRs that also modify the `batch_queue.go` file.